### DISCUSSION TITLE
ORM creation very slow when used with asyncio.gather

### DISCUSSION THREAD

**Discussion by massi82** [sqlalchemy/sqlalchemy#10808] [Usage Questions] [answered]
In my web app I'm using PostgreSQL (via asyncpg connector) with FastAPI and I'm having serious performance issues when I'm trying to create records via asyncio.gather. It's not easy to create a simple script showing the issue, but more or less it can be summarized like this:

```
# Database setup
engine = create_async_engine(DATABASE_URL, future=True, echo=False, pool_size=30, max_overflow=50)
async_session = sessionmaker(engine, expire_on_commit=False, class_=AsyncSession)

class Order(Base):
    __tablename__ = 'order'
    id = Column(pgUUID(as_uuid=True), primary_key=True, default=uuid4, unique=True, nullable=False)
    user_id = Column(String, ForeignKey('user.id'), nullable=False)
   # other columns

class OrderDAL(BaseDAL):
    def __init__(self, session: AsyncSession):
        super().__init__(session)

    @log_execution_time
    async def create_orders(self, order_data):
        try:
            new_order = Order(**order_data)
            self.session.add(new_order)
            await self.session.flush()
            return new_order, None
        except SQLAlchemyError as e:
            await self.session.rollback() 
            raise e  

@asynccontextmanager
async def order_dal():
    async with async_session() as session:
        async with session.begin():
            yield OrderDAL(session)

async def main():
    # Main execution block
    async with order_dal() as dal:
        orders_list = [...]
        tasks = [order_dal.create_orders(order_data) for order_data in orders_list]
        responses = await asyncio.gather(*tasks)

if __name__ == "__main__":
    asyncio.run(main())

```

What happens is that even if I create only 2 tasks, the first one takes around 0.1 seconds (as I expect), the second goes immediately up to 1.3 seconds. If I increase the number of tasks the time goes up accordingly. The point where the code gets stuck is the flush in the function create_orders. Any suggestion about what I'm doing wrong?

**Comment by zzzeek** [comment]
well you are sharing a single ORM `Session` between concurrent tasks which is not allowed and will generally have disastrous results, as it not only shares the non-concurrency-safe `Session` between tasks but also a single database connection.  if you want to do concurrent tasks you'd need to use distinct sessions.

see https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#using-asyncsession-with-concurrent-tasks and https://docs.sqlalchemy.org/en/20/orm/session_basics.html#session-faq-threadsafe

**Reply by CaselIT** [reply]
It also make very little sense to use gather there from a performance point of view.

If you create a list of orders, add them all to the session and flush a single multi row insert is usually used that's faster than many single row inserts

**Reply by massi82** [reply]
Thanks a lot for your answers, starting from the documentation I have created a simple script representing the issue I am experiencing:

```
from sqlalchemy.orm import declarative_base
from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession
from sqlalchemy.orm import sessionmaker
from contextlib import asynccontextmanager
from sqlalchemy import Date, Integer, Column, Enum, String, Float, Text
from sqlalchemy.dialects.postgresql import UUID as pgUUID
from uuid import uuid4
import asyncio
import time
import random
import string

Base = declarative_base()
db_url = "postgresql+asyncpg://postgres:password@localhost:5434/mydb"
engine = create_async_engine(db_url, future=True, echo=False, pool_size=30, max_overflow=50)
async_session = sessionmaker(engine, expire_on_commit=False, class_=AsyncSession)

class Order(Base):
    __tablename__ = 'order'
 
    id = Column(pgUUID(as_uuid=True), primary_key=True, default=uuid4, unique=True, nullable=False)
    field1 = Column(Date, nullable=True)
    field2 = Column(Date, nullable=True)
    field3 = Column(Float, default=0) 
    field4 = Column(Float, default=0)
    field5 = Column(Float, default=0)
    field6 = Column(Float, default=0)
    field7 = Column(Float, default=0)
    field8 = Column(Float, default=0)
    field9 = Column(String, nullable=True)
    field10 = Column(String, nullable=True)
    field11 = Column(String, nullable=True)
    field12 = Column(Text, nullable=True) 
    field13 = Column(Integer, default=0) 
    field14 = Column(String, nullable=False)

class OrderDAL():
    def __init__(self, session: AsyncSession):
        self.session = session
         
    async def create_orders(self, orders_data):         
        new_orders = []
        for order_data in orders_data:
            new_order = Order(**order_data)
            self.session.add(new_order)
            new_orders.append(new_order)
        start = time.time()
        await self.session.flush()
        print("Flush finished in %.4f" %(time.time()-start))
        return new_orders, None

@asynccontextmanager
async def order_dal():
    async with async_session() as session:
        async with session.begin():
            yield OrderDAL(session)

async def create_order():    
    async with order_dal() as dal:  
        order = {
            "field3":random.uniform(800, 3000),
            "field4":random.uniform(700, 900),
            "field9": ''.join(random.choices(string.ascii_letters, k=5)),
            "field10": ''.join(random.choices(string.ascii_letters, k=5)),
            "field12": ''.join(random.choices(string.ascii_letters, k=5)),
            "field13": int(random.uniform(800, 3000)),
            "field14":''.join(random.choices(string.ascii_letters, k=5)),
        }
        created_orders, msg = await dal.create_orders([order]) 

N = 100

async def with_gather():
    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.drop_all)
        await conn.run_sync(Base.metadata.create_all)
    await setup_database()
    start = time.time()
    tasks = [create_order() for _ in range(N)]
    await asyncio.gather(*tasks)
    print("Exec time with asyncio.gather: %.4f" %(time.time()-start))

async def main():
    await with_gather()

if __name__ == "__main__":
    asyncio.run(main())
```

This is the output I get from the execution with N = 100:

```
Flush finished in 0.1026
Flush finished in 2.1762
Flush finished in 2.3335
...
Flush finished in 3.0469
Flush finished in 3.0469
Exec time with asyncio.gather: 3.1520
```

So I still don't get how the concurrent execution works. As you can see the first flush is done in about 0.1 seconds, that is the amount of time I expect from the query execution. The following ones immediately rise up to 2 seconds, getting up to 3 seconds for the last ones. This behavior is not clear to me, since I would instead expect that the first executions would require around 0.1 seconds, while  after some tens it might become reasonable to see longer times due to sessions competing (and waiting) for the available connections. 

Sorry if this is the expected behavior but I'm pretty new to the use of SQLAlchemy in an async context. 

Thanks in advance for the help!

**Reply by zzzeek** [reply, marked as answer]
the issue locally is that your create_all() uses a new connection from the pool, so the pool is pre-warmed with a single connection.  the rest of the operations, since they are concurrent, have to get a new connection from the pool each time, hence there's a wait.

you can pre-warm the pool like this:

```py
    warmup = [await engine.connect() for i in range(30)]
    for conn in warmup:
        await conn.close()

```

note the program will dump a lot of errors unless you use engine.dispose() at the end.

On that subject I've identified a critical issue in the async version of engine.dispose() that's a separate issue from this one.

**Reply by massi82** [reply]
Ah, got it, thank you!
